### PR TITLE
Switch to an explicit stack to avoid stack overflow when computing tree equivalence

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2018744

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs [hunk 1]
```diff
@@ -100,130 +101,147 @@ private static bool AreTokensEquivalent(GreenNode? before, GreenNode? after, Fun
 
         private static bool AreEquivalentRecursive(GreenNode? before, GreenNode? after, Func<SyntaxKind, bool>? ignoreChildNode, bool topLevel)
         {
-            if (before == after)
-            {
-                return true;
-            }
+            var stack = ArrayBuilder<(GreenNode? before, GreenNode? after)>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs:104]: view with whitespace off.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler ptal.

**Review by arkalyanms — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jaredpar  @RikkiGibson ptal.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs [hunk 2]
```diff
@@ -100,130 +101,147 @@ private static bool AreTokensEquivalent(GreenNode? before, GreenNode? after, Fun
 
         private static bool AreEquivalentRecursive(GreenNode? before, GreenNode? after, Func<SyntaxKind, bool>? ignoreChildNode, bool topLevel)
         {
-            if (before == after)
-            {
-                return true;
-            }
+            // Use an explicit stack so we can walk down deep trees without blowing the real stack.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs:104]: >Use an explicit stack [](http://example.com/codeflow?start=15&length=21)

Could we add a test that would previously fail? (Say, two trees with deeply nested binary operators, either equivalent or distinct at the deepest token.)

**Comment by CyrusNajmabadi** [comment]
@cston test (which failed before) added.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs:104]: done.

**Review by cston — APPROVED** [review]

**Pull request merged**